### ISSUE TITLE
[temp.spec.partial] Should template parameter pack in class template partial specialization that is not otherwise deduced be ill-formed?

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#314]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

Consider this example:
````cpp
template<class ...Args>
struct A{};

template<class F, class...Args>
struct A<F>{};
````

[Implementations](https://godbolt.org/z/Krhjrh5oK) render ill-formed for this case. However, we say template paramter pack can accept zero arguments, and [temp.fct.spec#temp.arg.explicit-note-1] explicitly says
> [Note 1: A trailing template parameter pack ([temp.variadic]) **not otherwise deduced will be deduced as an empty sequence of template arguments**. — end note]  

That is , in the above example, the deduction for the template parameter pack should behave the same as the following:
````cpp
template<class F, class...Args>
void fun(F){}
````

Moreover, the above example seems does not violate any rule in [temp.spec.partial].

**Comment by languagelawyer** [comment]
http://eel.is/c++draft/temp.spec.partial.match#3?

**Comment by xmh0511** [comment]
> http://eel.is/c++draft/temp.spec.partial.match#3?

However, `Args...` can be deduced to be an empty sequence of template arguments

**Comment by jensmaurer** [comment]
There is a general rule somewhere that a template is IFNDR if it is valid onl when a pack has no elements, I think.

**Comment by languagelawyer** [comment]
@jensmaurer but the specialization does not have anything like `static_assert(sizeof...(Args) == 0)` inside

**Comment by jensmaurer** [comment]
That's not the point. See [temp.res.general] p6.3.

**Comment by languagelawyer** [comment]
I know; how did «only empty is possible» become equal to «requires an empty»?

**Comment by jensmaurer** [comment]
A partial specialization is a template in its own right (although not a primary template).  I think there are no valid (full) specializations of the partial specialization with a non-empty pack, thus violating "every valid specialization of a variadic template requires an empty template parameter pack".

**Comment by languagelawyer** [comment]
> A partial specialization is a template in its own right (although not a primary template)

Right. So, from [temp.res.general] p6.3, the primary template does not interfere when substituting template params. We get (invent?) some (unique?) types, and look what would happen if we imaginary replaced a pack with them.

> I think there are no valid (full) specializations of the partial specialization with a non-empty pack

What exactly will be invalid if one substitutes non-empty `Args` in the partial specialization?

**Comment by jensmaurer** [comment]
> What exactly will be invalid if one substitutes non-empty Args in the partial specialization?

It's not a possible specialization to start with, because it cannot ever happen given the primary template. We could add a clarification in [temp.res.general] p6.3.

**Comment by xmh0511** [comment]
@jensmaurer If [temp.res.general] p6.3. would apply to the partial specialization(a template), it should also apply to this case
````cpp
template<class F, class...Args>
void fun(F){}
````
The template specialization would otherwise be ill-formed because the deduction failture except that `Args` is empty.

**Comment by jensmaurer** [comment]
Yes, it does.  It's IFNDR, so your compiler might not deliver a diagnostic.

**Comment by jensmaurer** [comment]
You're right, sorry, the function template example is fine, because you can give explicit template arguments.  (You don't have to rely on deduction, unlike the partial specialization case.)

**Comment by RealLitb** [comment]
I think this touches on several existing issue reports. One is https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#697 (which could just forbid defaulting the pack to be empty.. just as explicit default template arguments are forbidden on partial specializations aswell). It's clear that not everything in in arg.explicit applies to partial specializations. For example:

> If all of the template arguments can be deduced, they may all be omitted; in this case, the empty template argument list <> itself may also be omitted.

Another issue is https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#549 which is slightly off with Jens' (and, honestly, mine) interpretation, because there is no valid generated specialization for the the shown partial specialization, and it is not instantiated, so I would have thought it is IFNDR. But nevertheless, Issue #549 proposed a solution for that, and for this issue!

> Each template-parameter shall appear at least once in the template-id outside a non-deduced context.

However, I can't find that text in recent drafts anymore, even though it says "[Adopted at the February, 2016 meeting.]". Can anyone shed some light on that?
